### PR TITLE
Tutorial for Lagrangian Neural Network

### PR DESCRIPTION
**Pull request by a-b-h-a-y-s-h-i-n-d-e**
## Description

Tutorial for LNN, should be ready for merged once deepchem LNN model gets merged.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ x ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas
- [ x ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ x ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb [hunk 1]
```diff
@@ -0,0 +1,728 @@
+{
```
Comment (sudo-rsingh) [inline examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: <p>There is a keyboard interrupt, can you pls fix it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/4561/discussion/#comment-2511842603'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (sudo-rsingh) [inline examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: <p><code>UserWarning: Converting a tensor with requires_grad=True to a scalar may lead to unexpected behavior.</code></p><p>There is a warning informing about detach not being used can you fix it. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/4561/discussion/#comment-2511842617'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (sudo-rsingh) [inline examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: <p>We usually keep this sections title as <code>Reference</code>.</p><p>And can you follow the same citation style which is in other older tutorials.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/4561/discussion/#comment-2511842626'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (sudo-rsingh) [inline examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: <p>Pls put the full url</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/4561/discussion/#comment-2511842633'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: There are tutorials which use `Reference` as title ( attached screenshot ), also I tried to keep similar with HNN tutorial, but if you want I can change that too

Screenshot From 2025-11-12 12-02-51

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: done

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: done

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: done

File: examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
```
Comment (JoseAntonioSiguenza) [inline examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: <p>please add citations</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/4561/discussion/#comment-2518378528'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (JoseAntonioSiguenza) [inline examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    from scipy.integrate import odeint
</pre><p>Please check <code>deepchem.utils</code> module. I believe DeepChem provides a utility function for solving differential equations using numerical methods.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/4561/discussion/#comment-2518378542'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: The repo from author has also used odeint thats why I decided to use it. [link](https://github.com/MilesCranmer/lagrangian_nns)
also, `deepchem.utils` particulary to integrate systems dont offer odeint ( based on LSODA ) , it has rk4 methods.

The difference between 2 function is that, `rk4` methods follows fix step size along with calculating 4 sub-steps , whereas `odeint` chooses step size automatically and also it chooses algorithm based on equation ( adams method / BDF method )


TLDR;
`rk4` methods are useful when the ODEs are simple, but in case of LNNs we have non-linear equations to simualte thats the reason to use `odeint`

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: done

Comment (sudo-rsingh) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: Can you as part of this PR, implement a simple adaptive solver in DeepChem

Comment (a-b-h-a-y-s-h-i-n-d-e) [reply examples/tutorials/Introduction_to_Lagrangian_Neural_Network.ipynb:1]: I guess, implementing RK45 will be good option in deepchem, but I think, implementing that in separate PR will be good!

**Review by sudo-rsingh — APPROVED** [review]
LGTM

cc: @rbharath 

- I think adding a deepchem native adaptive solver in deepchem will be a better idea as we have related tooling.

**Review by rbharath — APPROVED** [review]
LGTM
